Extract function optimizations, ~20% speedup

nol13 opened the pull request:
Did this separately from all of my other changes, though a bit of related code. Added parameter "process_s1" to most of the scorers that is used by extractWithoutOrder to tell them not to run full_process on the query again, so that it doesn't get run on it for each choice. Does add an extra parameter that probably doesn't need to be exposed publicly though, so maybe a better way to do that.

Added a couple additional benchmarks as well, was getting about ~20% improvement for me depending on the scorer used.

worldofnick commented:
@rajseatgeek can we get this in

josegonzalez commented:
Lol @rajseatgeek aren't you still team :droid:?

@worldofnick have you confirmed that all tests pass with this merged into master? I must have missed this due to the holidays last year...

rajseatgeek commented:
@josegonzalez i thought i was but hey, i'll merge this if you want me to

nol13 commented:
^ just small change to make benchmark i added fully deterministic, was considering adding this last time i looked at this i think. makes no meaningful difference, but still..

(also updated pypy version in travis.yml since i guess travis no longer supports <=3.2 apparently)
